tcc/src/main/java/com/alibaba/fescar/rm/tcc/TCCResourceManager.java

```diff
@@ -0,0 +1,171 @@
+
+import com.alibaba.fastjson.JSON;
+import com.alibaba.fescar.common.Constants;
+import com.alibaba.fescar.common.exception.FrameworkException;
+import com.alibaba.fescar.common.exception.ShouldNeverHappenException;
+import com.alibaba.fescar.common.util.StringUtils;
+import com.alibaba.fescar.core.exception.TransactionException;
+import com.alibaba.fescar.core.model.BranchStatus;
+import com.alibaba.fescar.core.model.BranchType;
+import com.alibaba.fescar.core.model.Resource;
+import com.alibaba.fescar.rm.AbstractResourceManager;
+import com.alibaba.fescar.rm.tcc.api.BusinessActionContext;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * TCC resource manager
+ *
+ * @author zhangsen
+ */
+public class TCCResourceManager extends AbstractResourceManager {
+
+	/**
+	 * TCC resource cache
+	 */
+	private Map<String, Resource> tccResourceCache = new ConcurrentHashMap<String, Resource>();
+
+    /**
+     * Instantiates a new Tcc resource manager.
+     */
+    public TCCResourceManager(){
+	}
+
+	/**
+	 * registry TCC resource
+	 * @param resource The resource to be managed.
+	 */
+	@Override
+	public void registerResource(Resource resource) {
+		TCCResource tccResource = (TCCResource) resource;
+		tccResourceCache.put(tccResource.getResourceId(), tccResource);
+		super.registerResource(tccResource);
+	}
+
+	@Override
+	public Map<String, Resource> getManagedResources() {
+		return tccResourceCache;
+	}
+
+	/**
+	 * TCC branch commit
+	 * @param branchType
+	 * @param xid             Transaction id.
+	 * @param branchId        Branch id.
+	 * @param resourceId      Resource id.
+	 * @param applicationData Application data bind with this branch.
+	 * @return
+	 * @throws TransactionException
+	 */
+	@Override
+	public BranchStatus branchCommit(BranchType branchType, String xid, long branchId, String resourceId, String applicationData) throws TransactionException {
+		TCCResource tccResource = (TCCResource) tccResourceCache.get(resourceId);
+		if(tccResource == null){
+			throw new ShouldNeverHappenException("TCC resource is not exist, resourceId:" + resourceId);
+		}
+		Object targetTCCBean = tccResource.getTargetBean();
+		Method commitMethod = tccResource.getCommitMethod();
+		if(targetTCCBean == null || commitMethod == null){
+			throw new ShouldNeverHappenException("TCC resource is not available, resourceId:" + resourceId);
+		}
+		try {
+			boolean result = false;
+			//BusinessActionContext
+			BusinessActionContext businessActionContext = getBusinessActionContext(xid, branchId, resourceId, applicationData);
+			Object ret = commitMethod.invoke(targetTCCBean, businessActionContext);
+			if(ret != null && ret instanceof TwoPhaseResult){
+				result = ((TwoPhaseResult)ret).isSuccess();
+			}else {
+				result = (boolean) ret;
+			}
+			return result ? BranchStatus.PhaseTwo_Committed:BranchStatus.PhaseTwo_CommitFailed_Retryable;
```

Comment:
when result type is TwoPhaseResult and ret.isSuccess()=false , just return PhaseTwo_CommitFailed_Retryable, why not print TwoPhaseResult#msg to check problem？